### PR TITLE
Refactor metrics controller tests to fix flaky behaviour

### LINKED ISSUE #4391: Flaky Unit Test: TestControllerGameServersNodeState
I reckon this is because of the upgrade to Go 1.25, so now this is a difference in behaviour..

* https://console.cloud.google.com/cloud-build/builds/4215f885-e5c9-4d1f-80a6-926a4fd1d1c5;step=19?project=agones-images
* https://github.com/googleforgames/agones/pull/4388
```
--- FAIL: TestControllerGameServersNodeState (0.10s)
    controller_test.go:88: 
        	Error Trace:	/go/src/agones.dev/agones/pkg/metrics/controller_test.go:88
        	            				/go/src/agones.dev/agones/pkg/metrics/controller_test.go:605
        	Error:      	Not equal: 
        	            	expected: &metricdata.Distribution{Count:3, Sum:3, SumOfSquaredDeviation:2, BucketOptions:(*metricdata.BucketOptions)(0xc000d869a8), Buckets:[]metricdata.Bucket{metricdata.Bucket{Count:1, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:1, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:1, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}}}
        	            	actual  : &metricdata.Distribution{Count:1, Sum:1, SumOfSquaredDeviation:0, BucketOptions:(*metricdata.BucketOptions)(0xc000d868e8), Buckets:[]metricdata.Bucket{metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:1, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}, metricdata.Bucket{Count:0, Exemplar:(*metricdata.Exemplar)(nil)}}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,5 +1,5 @@
        	            	 (*metricdata.Distribution)({
        	            	- Count: (int64) 3,
        	            	- Sum: (float64) 3,
        	            	- SumOfSquaredDeviation: (float64) 2,
        	            	+ Count: (int64) 1,
        	            	+ Sum: (float64) 1,
        	            	+ SumOfSquaredDeviation: (float64) 0,
        	            	  BucketOptions: (*metricdata.BucketOptions)({
        	            	@@ -37,7 +37,3 @@
        	            	   (metricdata.Bucket) {
        	            	-   Count: (int64) 1,
        	            	-   Exemplar: (*metricdata.Exemplar)(<nil>)
        	            	-  },
        	            	-  (metricdata.Bucket) {
        	            	-   Count: (int64) 1,
        	            	+   Count: (int64) 0,
        	            	    Exemplar: (*metricdata.Exemplar)(<nil>)
        	            	@@ -147,2 +143,6 @@
        	            	    Exemplar: (*metricdata.Exemplar)(<nil>)
        	            	+  },
        	            	+  (metricdata.Bucket) {
        	            	+   Count: (int64) 0,
        	            	+   Exemplar: (*metricdata.Exemplar)(<nil>)
        	            	   }
        	Test:       	TestControllerGameServersNodeState
        	Messages:   	metric: gameservers_node_count, tags: [], values don't match; got: &{1 1 0 0xc000d868e8 [{0 <nil>} {1 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>}]}, want: &{3 3 2 0xc000d869a8 [{1 <nil>} {1 <nil>} {1 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>} {0 <nil>}]}
{"message":"Wait for cache sync","severity":"info","time":"2025-12-18T02:35:08.006298316Z"}
{"message":"unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:\"\", Verb:\"watch\", Resource:schema.GroupVersionResource{Group:\"agones.dev\", Version:\"v1\", Resource:\"gameserversets\"}, Subresource:\"\"}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:\"\"}, ListOptions:v1.ListOptions{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, LabelSelector:\"\", FieldSelector:\"\", Watch:true, AllowWatchBookmarks:true, ResourceVersion:\"\", ResourceVersionMatch:\"\", TimeoutSeconds:(*int64)(0xc0006c50f8), Limit:0, Continue:\"\", SendInitialEvents:(*bool)(nil)}}","severity":"error","time":"2025-12-18T02:35:08.007016927Z"}
{"message":"unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:\"\", Verb:\"watch\", Resource:schema.GroupVersionResource{Group:\"agones.dev\", Version:\"v1\", Resource:\"gameserversets\"}, Subresource:\"\"}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:\"\"}, ListOptions:v1.ListOptions{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, LabelSelector:\"\", FieldSelector:\"\", Watch:true, AllowWatchBookmarks:true, ResourceVersion:\"\", ResourceVersionMatch:\"\", TimeoutSeconds:(*int64)(0xc000a03968), Limit:0, Continue:\"\", SendInitialEvents:(*bool)(nil)}}","severity":"error","time":"2025-12-18T02:35:09.595846674Z"}
{"message":"unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:\"\", Verb:\"watch\", Resource:schema.GroupVersionResource{Group:\"agones.dev\", Version:\"v1\", Resource:\"gameserversets\"}, Subresource:\"\"}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:\"\"}, ListOptions:v1.ListOptions{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, LabelSelector:\"\", FieldSelector:\"\", Watch:true, AllowWatchBookmarks:true, ResourceVersion:\"\", ResourceVersionMatch:\"\", TimeoutSeconds:(*int64)(0xc000f98198), Limit:0, Continue:\"\", SendInitialEvents:(*bool)(nil)}}","severity":"error","time":"2025-12-18T02:35:11.240893779Z"}
{"message":"Wait for cache sync","severity":"info","time":"2025-12-18T02:35:13.117241643Z"}
{"message":"unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:\"\", Verb:\"watch\", Resource:schema.GroupVersionResource{Group:\"agones.dev\", Version:\"v1\", Resource:\"gameserversets\"}, Subresource:\"\"}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:\"\"}, ListOptions:v1.ListOptions{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, LabelSelector:\"\", FieldSelector:\"\", Watch:true, AllowWatchBookmarks:true, ResourceVersion:\"\", ResourceVersionMatch:\"\", TimeoutSeconds:(*int64)(0xc000012a38), Limit:0, Continue:\"\", SendInitialEvents:(*bool)(nil)}}","severity":"error","time":"2025-12-18T02:35:13.117728213Z"}
{"message":"Starting http server...","severity":"info","test":"TestMetrics_Endpoint_ExposesAllMetrics","time":"2025-12-18T02:35:14.221317353Z"}
{"message":"unhandled watch: testing.WatchActionImpl{ActionImpl:testing.ActionImpl{Namespace:\"\", Verb:\"watch\", Resource:schema.GroupVersionResource{Group:\"agones.dev\", Version:\"v1\", Resource:\"gameserversets\"}, Subresource:\"\"}, WatchRestrictions:testing.WatchRestrictions{Labels:labels.internalSelector(nil), Fields:fields.andTerm{}, ResourceVersion:\"\"}, ListOptions:v1.ListOptions{TypeMeta:v1.TypeMeta{Kind:\"\", APIVersion:\"\"}, LabelSelector:\"\", FieldSelector:\"\", Watch:true, AllowWatchBookmarks:true, ResourceVersion:\"\", ResourceVersionMatch:\"\", TimeoutSeconds:(*int64)(0xc0008efe48), Limit:0, Continue:\"\", SendInitialEvents:(*bool)(nil)}}","severity":"error","time":"2025-12-18T02:35:14.423247844Z"}
{"error":"http: Server closed","message":"http server closed","severity":"info","test":"TestMetrics_Endpoint_ExposesAllMetrics","time":"2025-12-18T02:35:14.545124258Z"}
FAIL
```

### PR DESCRIPTION
**Pull request by markmandel**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug

/kind cleanup

> /kind documentation
> /kind feature
> /kind hotfix
> /kind release

**What this PR does / Why we need it**:

Moved testing that is polls through Eventually() tests into the metric assert function - since all metric data is eventually consistent.

Also centralizes edge cases around Distribution metrics.

**Which issue(s) this PR fixes**:

Closes #4391

**Special notes for your reviewer**:

N/A

### CODE REVIEW

File: pkg/metrics/controller_test.go [hunk 1]
```diff
@@ -59,34 +58,85 @@ type expectedMetricData struct {
...
+			c.lock.Lock()
+			reader.ReadAndExport(exporter)
+			c.lock.Unlock()
+			for _, m := range exporter.metrics {
+				if m.Descriptor.Name == metricName {
+					return true
+				}
+			}
+			return false
+		}, 10*time.Second, 100*time.Millisecond)
```
Comment (markmandel) [inline pkg/metrics/controller_test.go:91]: I am not sure why this fixes things, but basically just copying the structure of other working tests. OpenCensus can be weird.

**Comment by markmandel** [comment]
Oh yeah - I also ran all the tests in `pkg/metrics/controller_test.go` 200 times, and it all passed.

File: pkg/metrics/controller_test.go (hunk 1 above)
Comment (lacroixthomas) [reply pkg/metrics/controller_test.go:91]: I wonder if it was not because of a race condition, and that the lock / unlock around reader.ReadAndExport you added would be the patch ?🤔 There must be other things as well

**Review by lacroixthomas — APPROVED** [review]
LGTM !

File: pkg/metrics/controller_test.go (hunk 1 above)
Comment (markmandel) [reply pkg/metrics/controller_test.go:91]: I tried with just an eventually added to the final assert... and things did not go as planned. In theory it should have worked - but this seemed good enough 🤷🏻 it tested what we wanted to test.

The locking is to avoid data races collecting and then exporting metrics, since opencensus is not thread safe! It happened very rarely when I was running this lots.

Side note - didn't know `assert.EventuallyWithT` was a thing - I could refactor some tests to use it. Very handy.
